### PR TITLE
[Feature] ViT Full CUDA Graph

### PR DESCRIPTION
**Pull request by b-mu**
## Purpose

Add full CUDA graph for the ViT to reduce kernel launch overheads.

**Features:**
  - **Budget-based graphs with a maximum batch size**:
      - Capture CUDA graphs at configurable token budgets (e.g., `[256, 512, 1024, 2048, 4096]`).
      - Pad sequence metadata (e.g. cu_seqlen) so that we can use the same budget-based graph for various number of images
  during replays.
  - **Greedy bin-packing**:
      - Sort images in a batch in ascending order to reduce the number of graphs.
  - **Data-parallel (DP) support**:
      - When `mm_encoder_tp_mode=data`, each TP rank runs the ViT independently via data parallelism.
  - **FlashInfer cuDNN attention support**:
      - Override FlashInfer buckets in the CUDA graph path.
  - **Model-agnostic protocol**:
      - `SupportsEncoderCudaGraph` protocol in `interfaces.py` — models opt in by implementing 9 protocol methods for input
   handling, metadata computation, and forward dispatch.
      - `EncoderCudaGraphManager` is fully model-agnostic; all model-specific logic (grid config, dummy inputs, embedding
  computation) lives in the model class.

**New config flags** (via `--compilation-config`):
  - `cudagraph_mm_encoder: true` — enable encoder CUDA graph
  - `encoder_cudagraph_token_budgets: [...]` — list of token budget sizes to capture
  - `encoder_cudagraph_max_images_per_batch: N` — max images per graph replay

**Files changed:**
  - `vllm/config/compilation.py` — new config flags
  - `vllm/model_executor/models/interfaces.py` — `SupportsEncoderCudaGraph` protocol and `supports_encoder_cudagraph()`
  type guard
  - `vllm/model_executor/models/qwen3_vl.py` — implement `SupportsEncoderCudaGraph` on `Qwen3VLForConditionalGeneration`
  - `vllm/v1/worker/gpu/mm/encoder_cudagraph_defs.py` — `EncoderCudaGraphConfig`, `EncoderCudaGraphCaptureInputs`,
  `EncoderCudaGraphReplayBuffers` dataclasses
  - `vllm/v1/worker/gpu/mm/encoder_cudagraph.py` — `EncoderCudaGraphManager` (capture, replay, packing, DP)
  - `vllm/v1/worker/gpu_model_runner.py` — integration into V1 model runner
  - `tests/v1/cudagraph/test_encoder_cudagraph.py` — unit and GPU tests

cc @maxyanghu @wangshangsam @Anerudhan

## Test Plan

**Unit Tests:**
```bash
pytest tests/v1/cudagraph/test_encoder_cudagraph.py -v
```

**End-to-End Tests**:
- **Single GPU**: Qwen3-VL-30B-A3B-Instruct, VisionArena-Chat dataset, 3000 prompts + 300 warmup
```
vllm bench mm-processor \
  --model Qwen/Qwen3-VL-30B-A3B-Instruct \
  --dataset-name hf --dataset-path lmarena-ai/VisionArena-Chat \
  --num-prompts 3000 --num-warmups 300 \
  --max-model-len 32768 --seed 42 \
  --mm-encoder-attn-backend FLASHINFER \
  --compilation-config '{"cudagraph_mm_encoder": true, "encoder_cudagraph_token_budgets": [512, 1024, 1536, 2048, 2560, 3072, 3584, 4096, 4864], "encoder_cudagraph_max_images_per_batch": 8}'
```
- **Multi GPU**: Qwen3-VL-32B-Instruct, 4×GB200 TP=4 + ViT DP=4, random-mm dataset (20 imgs/req, 336×336), 1000 prompts + 200 warmup
```
vllm bench mm-processor \
  --model Qwen/Qwen3-VL-32B-Instruct \
  --dataset-name random-mm \
  --random-mm-base-items-per-request 20 \
  --random-mm-num-mm-items-range-ratio 0.0 \
  --random-mm-bucket-config '{"(336,336,1)": 1.0}' \
  --num-prompts 1000 --num-warmups 200 \
  --max-model-len 8192 --seed 42 \
  --mm-encoder-attn-backend FLASHINFER \
  --tensor-parallel-size 4 --mm-encoder-tp-mode data \
  --compilation-config '{"cudagraph_mm_encoder": true, "encoder_cudagraph_token_budgets": [512, 1024, 1536, 2048, 2560, 3072, 3584, 4096, 4864], "encoder_cudagraph_max_images_per_batch": 8}'
```

## Test Result

**Single GPU** (Qwen3-VL-30B, 1×GB200, VisionArena-Chat, 3000 prompts):

| Backend | Mean | P99 |
|---|---|---|
| FLASH_ATTN | +11.8% (5.13→4.52ms) | +31.6% (9.16→6.26ms) |
| FLASH_ATTN | +19.6% (5.42→4.36ms) | +40.3% (10.87→6.49ms) |

**Multi GPU** (Qwen3-VL-32B, 4×GB200 TP=4 DP=4, random-mm 20img/req, 1000 prompts):

| Backend | Mean | P99 |
|---|---|---|
| FLASH_ATTN | +18.4% (28.39→23.16ms) | +14.0% (238.78→205.28ms) |
| FLASHINFER | +44.4% (23.24→12.91ms) | +84.9% (172.41→26.05ms) |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -2494,7 +2498,22 @@ def _execute_mm_encoder(
                 with self.timed_encoder_operation(
                     should_time, mm_lora_refs, current_item_idx, num_items
                 ):
-                    curr_group_outputs = model.embed_multimodal(**mm_kwargs_group)
+                    cudagraph_output = None
+                    if (
+                        self.encoder_cudagraph_manager is not None
+                        and modality == "image"
+                        and "pixel_values" in mm_kwargs_group
+                        and "image_grid_thw" in mm_kwargs_group
```
Comment (Isotr0py) [inline vllm/v1/worker/gpu_model_runner.py]: Hmmm, I feel this is too model-specific, and it 's difficult to use for other models with different mm_kwargs naming.

Can we execute `encoder_cudagraph_manage` with mm_kwargs_group directly?

File: vllm/v1/worker/gpu/mm/encoder_cudagraph.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+        # example configuration that produces token_budget tokens.
+        # At runtime, actual images will be packed in any
+        # combination that fits the budget.
+        dummy_grid_config = self._generate_grid_config_for_budget(
+            token_budget, self.max_batch_size
+        )
+
+        dummy_pixel_values, dummy_grid_thw = self._prepare_dummy_inputs(
+            dummy_grid_config
+        )
```
Comment (Isotr0py) [inline vllm/v1/worker/gpu/mm/encoder_cudagraph.py]: I feel this is something belongs to `DummyInputsBuilder`, otherwise the CG manager's dummy data creation could be quite complicated if we want to support other models.

**Review by Isotr0py — COMMENTED** [review]
With a first glance, my concern is the generality of current encoder CG manager. I feel current implementation is too qwen3vl-specific (MRoPE + ViT RoPE), and it's difficult to boardcast this CG support to other ViTs.

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (b-mu) [reply vllm/v1/worker/gpu_model_runner.py]: To address the general concern about encoder cudagraph manager being too model specific, I'm thinking about having a class `SupportsEncoderCudaGraph(Protocol)` in `vllm/model_executor/models/interfaces.py`. To use encoder cudagraph support for a model, the model needs to implement a list of methods to tell the encoder cudagraph manager how to extract inputs, sequence metadata, etc. The encoder cudagraph manager would be model agnostic. The specific concern here in gpu_model_runner.py can use `mm_kwargs_group` directly. What do you think?

Comment (Isotr0py) [reply vllm/v1/worker/gpu_model_runner.py]: Sounds great.

Comment (b-mu) [reply vllm/v1/worker/gpu_model_runner.py]: We introduced `SupportsEncoderCudaGraph` protocol in `interfaces.py` with 9 protocol methods. The manager is now model-agnostic — all Qwen3-VL-specific logic lives in `qwen3_vl.py` implementing the protocol.

The specific concern here in `gpu_model_runner.py` now uses `mm_kwargs_batch` directly and uses `self.encoder_cudagraph_manager.supports_modality(modality)` instead of checking for specific keys.

File: vllm/v1/worker/gpu/mm/encoder_cudagraph.py (hunk 2 above)
Comment (b-mu) [reply vllm/v1/worker/gpu/mm/encoder_cudagraph.py]: Dummy input generation has been moved out of the manager and into the protocol method `prepare_encoder_cudagraph_capture_inputs()`. Each model implements its own dummy input logic (e.g., in qwen3_vl.py).

File: vllm/model_executor/models/qwen3_vl.py [hunk 3]
```diff
@@ -531,44 +532,55 @@ def forward(
         self,
         x: torch.Tensor,
         grid_thw: torch.Tensor | list[list[int]],
+        *,
+        encoder_metadata: dict[str, torch.Tensor] | None = None,
```
Comment (Isotr0py) [inline vllm/model_executor/models/qwen3_vl.py:621]: I think we can add a `prepare_encoder_metadata` method to avoid duplicated implementation:
```python3
def prepare_encoder_metadata(...) -> dict[str, Any]:
    ...
    max_seqlen = torch.tensor(
        MMEncoderAttention.compute_max_seqlen(self.attn_backend, cu_seqlens),
        dtype=torch.int32,
        device=self.device,
    )
    cu_seqlens = MMEncoderAttention.maybe_recompute_cu_seqlens(
        self.attn_backend,
        cu_seqlens,
        self.hidden_size,
        self.tp_size,
    )
    cu_seqlens = torch.from_numpy(cu_seqlens).to(self.device, non_blocking=True)
    return {"max_seq_len": ...}

def forward(
    self,
    x: torch.Tensor,
    grid_thw: torch.Tensor | list[list[int]],
    *,
    encoder_metadata: dict[str, torch.Tensor] | None = None,
):
    ...
    if encoder_metadata is None:
        encoder_metadata = self.prepare_encoder_metadata(...)
    ...
    for layer_num, blk in enumerate(self.blocks):
        hidden_states = blk(
            hidden_states,
            **encoder_metadata,
        )
```
Then we can reuse it in `prepare_encoder_cudagraph_capture_inputs` and `prepare_encoder_cudagraph_replay_buffers`:
```python3
def prepare_encoder_cudagraph_replay_buffers(
        self,
        mm_kwargs: dict[str, Any],
        max_batch_size: int,
    ):
        from vllm.v1.worker.gpu.mm.encoder_cudagraph_defs import (
            EncoderCudaGraphReplayBuffers,
        )
        buffers = self.visual.prepare_encoder_metadata(...)
        ...
        return EncoderCudaGraphReplayBuffers(buffers=buffers)
```

WDYT?

Comment (b-mu) [reply vllm/model_executor/models/qwen3_vl.py:621]: good catch, I have added the helper `prepare_encoder_metadata()` and used it for eager forward pass, cudagraph capture/replay for qwen3 vl.

File: vllm/v1/worker/gpu/mm/encoder_cudagraph.py [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+            max_output_tokens_per_rank: Max output tokens across all ranks
+                (for padding during all_gather).
+        """
+        from vllm.distributed import (
+            get_tensor_model_parallel_rank,
+            get_tensor_model_parallel_world_size,
+        )
+        from vllm.model_executor.models.vision import (
+            get_load_balance_assignment,
+        )
```
Comment (Isotr0py) [inline vllm/v1/worker/gpu/mm/encoder_cudagraph.py]: I think we can move it to top imports.

File: vllm/v1/worker/gpu/mm/encoder_cudagraph.py [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+        images_per_rank: list[int],
+        max_output_tokens_per_rank: int,
+    ) -> list[torch.Tensor]:
+        """Gather outputs from all TP ranks and reorder to original sequence.
+
+        Assumes 2D output tensors [tokens, hidden]. Follows the same
+        pad -> all_gather -> unpad -> reorder algorithm as
+        run_dp_sharded_mrope_vision_model() in the eager path.
+        """
+        from vllm.distributed import tensor_model_parallel_all_gather
```
Comment (Isotr0py) [inline vllm/v1/worker/gpu/mm/encoder_cudagraph.py]: Ditto.

File: vllm/v1/worker/gpu/mm/encoder_cudagraph.py
Comment (Isotr0py) [inline vllm/v1/worker/gpu/mm/encoder_cudagraph.py:1]: @ywang96 I remember `v1/worker/gpu` is reserved for v2 model runner, is it fine to place it here and only execute in v1 as well?

File: vllm/config/compilation.py [hunk 6]
```diff
@@ -489,6 +489,23 @@ class CompilationConfig:
...
+    """Enable CUDA graph capture for multimodal encoder (ViT).
+    When enabled, captures full encoder forward as CUDA graph
+    for each token budget level.
+    Requires encoder_cudagraph_token_budgets to be specified."""
+
+    encoder_cudagraph_token_budgets: list[int] = field(default_factory=list)
+    """Token budget levels for encoder CUDA graph capture.
+    Each budget defines a fixed token capacity. At runtime, images are greedy-packed
+    into the smallest fitting budget and the corresponding CUDA graph is replayed.
+    Example: [2048, 4096, 8192, 13824]"""
```
Comment (Isotr0py) [inline vllm/config/compilation.py:505]: I think asking user to pass `encoder_cudagraph_token_budgets` manually will cause poor user experience. Is it possible to automatically infer best `encoder_cudagraph_token_budgets` (at least cover most of possible image size) from `get_mm_max_tokens_per_item`?
https://github.com/vllm-project/vllm/blob/fad09e8a1f51b31eba1f42ff5d651256c77a734d/vllm/multimodal/encoder_budget.py#L15-L41

**Review by Isotr0py — APPROVED** [review]
As an experimental functionality, the model interface and cg manager overall LGTM now. But still have a minor suggestion about UX.

File: vllm/config/compilation.py (hunk 6 above)
Comment (b-mu) [reply vllm/config/compilation.py:505]: It seems that `get_mm_max_tokens_per_item` returns the theoretical maximum that the model can consume, which might not get hit by a dataset. Capturing a CUDA graph that's too large could waste GPU memory. On the other hand, if an image is large enough, the kernel launch overhead might not be that visible, so the price of falling back to eager is low. 

I'm thinking about adding a new method in the protocol to provide the `(min budget, max budget)` of that model. For qwen3-vl, I'm thinking about using `max_num_batched_tokens` as the maximum budget, and using `224x224 ~ 64` tokens as the minimum budget. What do you think?

File: vllm/v1/worker/gpu/mm/encoder_cudagraph.py (hunk 4 above)
Comment (b-mu) [reply vllm/v1/worker/gpu/mm/encoder_cudagraph.py]: done.

File: vllm/v1/worker/gpu/mm/encoder_cudagraph.py (hunk 5 above)
Comment (b-mu) [reply vllm/v1/worker/gpu/mm/encoder_cudagraph.py]: done.

File: vllm/config/compilation.py (hunk 6 above)
Comment (Isotr0py) [reply vllm/config/compilation.py:505]: > I'm thinking about using max_num_batched_tokens as the maximum budget, and using 224x224 ~ 64 tokens as the minimum budget.

Sounds great.
